Sources/Benchmarks/AnyValueBenchmarks.swift

```diff
@@ -0,0 +1,417 @@
+// Copyright 2020 Penguin Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Benchmark
+import PenguinStructures
+
+@_optimize(none)
+fileprivate func doNotOptimizeAway1(_ x: UnsafeRawPointer) {}
+
+@inline(__always)
+fileprivate func doNotOptimizeAway<T>(_ x: T) {
+  withUnsafePointer(to: x) { doNotOptimizeAway1($0) }
+}
+
+@_optimize(none)
+fileprivate func hideValue1<T>(_ x: UnsafeMutablePointer<T>) {}
+
+@inline(__always)
+fileprivate func hideValue<T>(_ x: T) -> T {
+  var copy = x
+  withUnsafeMutablePointer(to: &copy) { hideValue1($0) }
+  return copy
+}
+
+typealias TLS = PosixConcurrencyPlatform_.ThreadLocalStorage
+
+fileprivate class VTables {
+  var value: [Array2<TypeID>: UnsafeRawPointer] = [:]
+  init() {}
+}
+
+fileprivate let vTableKey = TLS.makeKey(for: Type<VTables>())
+
+@inline(never)
+fileprivate func makeVTables() -> VTables {
+  let r = VTables()
+  TLS.set(r, for: vTableKey)
+  return r
+}
+
+func pretendToMutate(_ x: inout UnsafeRawPointer) -> UnsafeRawPointer { x }
+
+fileprivate func demandVTable<Table>(
+  _ tableID: Array2<TypeID>, create: ()->Table
+) -> UnsafePointer<Table> {
+  
+  func allocateTable() -> UnsafeRawPointer {
+    let r = UnsafeMutablePointer<Table>.allocate(capacity: 1)
+    r.initialize(to: create())
+    return .init(r)
+  }
+  
+  let tables = TLS.get(vTableKey) ?? makeVTables()
+  let tableAddress = pretendToMutate(&tables.value[tableID, default: allocateTable()])
+  return tableAddress.assumingMemoryBound(to: Table.self)
+}
+
+internal protocol AnyValueProtocol {
+  /// Creates an instance that stores `x`.
+  ///
+  /// - Postcondition: where `a` is the created instance, `a.storedType == T.self`, and `a[T.self]`
+  ///   is equivalent to `x`.
+  init<T>(_ x: T)
+  
+  /// The type of the value stored in `self`.
+  var storedType: Any.Type { get }
+
+  /// Accesses the `T` stored in `self`.
+  ///
+  /// - Requires: `storedType == T.self`.
+  subscript<T>(_: Type<T>) -> T { get set }
+
+  /// Unsafely accesses the `T` stored in `self`.
+  ///
+  /// - Requires: `storedType == T.self`.
+  subscript<T>(unsafelyAssuming _: Type<T>) -> T { get set }
+
+  /// Stores `x` in `self`.
+  mutating func store<T>(_ x: T)
+
+  /// The stored value.
+  var asAny: Any { get }
+}
+
+extension AnyValue: AnyValueProtocol {}
+
+internal struct NaiveAnyBased_AnyValue: AnyValueProtocol {
```

Comment:
As described to me orally, I really like what you're doing (i.e. benchmarking a few different types of type erasure), but it's not so obvious when reading the code or looking at these types. Please do consider adding some brief context. Thanks!